module/state_synchronization/execution_data_requester.go

```diff
@@ -0,0 +1,548 @@
+
+	e := &ExecutionDataRequester{
+		log:                  log.With().Str("engine", "executiondata_requester").Logger(),
+		ds:                   datastore,
+		bs:                   blobservice,
+		eds:                  eds,
+		rootBlock:            rootBlock,
+		blocks:               blocks,
+		results:              results,
+		missingCheckInterval: missingBlockCheckInterval,
+		finalizedBlocks:      make(chan flow.Identifier),
+		configLoaded:         make(chan struct{}),
+	}
+
+	finalizationDistributor.AddOnBlockFinalizedConsumer(e.onBlockFinalized)
+
+	e.cm = component.NewComponentManagerBuilder().
+		AddWorker(e.finalizedBlockProcessor).
+		AddWorker(e.checkAndBackfill).
+		Build()
+	e.Component = e.cm
+
+	return e, nil
+}
+
+// onBlockFinalized is called when a new block is finalized, and passes the block ID to the
+// main worker for processing
+func (e *ExecutionDataRequester) onBlockFinalized(blockID flow.Identifier) {
+	if util.CheckClosed(e.cm.ShutdownSignal()) {
+		return
+	}
+	e.finalizedBlocks <- blockID
+}
+
+// finalizedBlockProcessor runs the main process that processes finalized block notifications and
+// requests ExecutionData for each block seal
+func (e *ExecutionDataRequester) finalizedBlockProcessor(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	e.last = lastProcessedHeight{ds: e.ds}
+	if err := e.last.load(ctx); err != nil {
+		e.log.Debug().Err(err).Msg("last processed height not found in db")
+		e.last.set(ctx, e.rootBlock.Header.Height)
+	}
+	e.log.Debug().Msgf("starting with last processed height %d", e.last.height())
+
+	e.missingBlocks = missingBlocksList{ds: e.ds}
+	if err := e.missingBlocks.load(ctx); err != nil {
+		e.log.Debug().Err(err).Msg("missing blocks list not found in db. using empty list")
+	}
+
+	close(e.configLoaded)
+
+	<-e.eds.Ready()
+	ready()
+
+	e.finalizationProcessingLoop(ctx)
+}
+
+// finalizationProcessingLoop watches for new finalized blocks then kicks off a goroutine to
+// process it
+// This uses a channel to receive the block instead of processing the finalized block directly in the
+// callback, because the irrecoverable context passed into start is only available via the worker
+func (e *ExecutionDataRequester) finalizationProcessingLoop(ctx irrecoverable.SignalerContext) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case blockID := <-e.finalizedBlocks:
+			go e.processFinalizedBlock(ctx, blockID)
+		}
+	}
+}
+
+// processFinalizedBlock downloads the ExecutionData for each of the blocks sealed in the finalized block
+func (e *ExecutionDataRequester) processFinalizedBlock(ctx irrecoverable.SignalerContext, blockID flow.Identifier) {
+	logger := e.log.With().Str("finalized_block_id", blockID.String()).Logger()
+	logger.Debug().Msg("received finalized block")
+
+	// get the block
+	block, err := e.blocks.ByID(blockID)
+	if errors.Is(err, storage.ErrNotFound) {
+		logger.Error().Err(err).Msg("finalized block is missing from protocol state db")
+		return
```

Comment:
I opted to make this component as non-intrusive as possible since it's not really required to run the node safely.

I'm fine crashing the node if the block should be in the store, but don't feel strongly either way.

In either case, since sealed blocks must be sequential, if we skip/miss any block, we'll detect it at the next block finalized block with seals, and queue up any heights we've missed.